File: backend/app/parsers/base_parser.py
"""
Base parser abstract class for multi-agent framework log parsing.

All framework-specific parsers (AutoGen, CrewAI, LangGraph, etc.) must
inherit from BaseParser and implement the parse() method.
"""

from abc import ABC, abstractmethod
from pathlib import Path
from typing import List
from app.models.types import Message


class BaseParser(ABC):
    """
    Abstract base class for all log file parsers.

    Each framework (AutoGen, CrewAI, etc.) has its own log format.
    This class defines the common interface for parsing logs into
    a standardized Message list.
    """

    def __init__(self):
        """Initialize the parser."""
        self.framework_name = self.__class__.__name__.replace("Parser", "").lower()

    @abstractmethod
    def parse(self, file_path: Path) -> List[Message]:
        """
        Parse a log file into a list of Message objects.

        Args:
            file_path: Path to the log file (typically JSONL or JSON)

        Returns:
            List of Message objects in chronological order

        Raises:
            ValueError: If the file format is invalid or cannot be parsed
            FileNotFoundError: If the file does not exist
        """
        pass

    def validate_file(self, file_path: Path) -> None:
        """
        Validate that the file exists and is readable.

        Args:
            file_path: Path to the file to validate

        Raises:
            FileNotFoundError: If file does not exist
            PermissionError: If file is not readable
        """
        if not file_path.exists():
            raise FileNotFoundError(f"Log file not found: {file_path}")

        if not file_path.is_file():
            raise ValueError(f"Path is not a file: {file_path}")

        if not file_path.stat().st_size > 0:
            raise ValueError(f"File is empty: {file_path}")

    def __str__(self) -> str:
        return f"{self.__class__.__name__}(framework={self.framework_name})"

    def __repr__(self) -> str:
        return self.__str__()


class ParserError(Exception):
    """Custom exception for parser-specific errors."""

    def __init__(self, message: str, line_number: int = None, original_error: Exception = None):
        self.message = message
        self.line_number = line_number
        self.original_error = original_error
        super().__init__(self.format_error())

    def format_error(self) -> str:
        """Format the error message with context."""
        error_msg = f"Parser Error: {self.message}"
        if self.line_number:
            error_msg += f" (line {self.line_number})"
        if self.original_error:
            error_msg += f" - Original error: {str(self.original_error)}"
        return error_msg

File: src/lib/parsers/base-parser.ts
/**
 * Base Parser - Abstract interface for framework-specific log parsers
 *
 * All parsers inherit from this class and implement the parse() method.
 */

import type { Message } from '@/lib/models/types';

/**
 * Custom error class for parser errors with line context.
 */
export class ParserError extends Error {
  lineNumber?: number;
  originalError?: Error;

  constructor(
    message: string,
    options?: { lineNumber?: number; originalError?: Error }
  ) {
    super(message);
    this.name = 'ParserError';
    this.lineNumber = options?.lineNumber;
    this.originalError = options?.originalError;
  }
}

/**
 * Abstract base class for log parsers.
 */
export abstract class BaseParser {
  abstract readonly frameworkName: string;

  /**
   * Parse a log file content into Message objects.
   *
   * @param content - The raw content of the log file
   * @returns List of Message objects in chronological order
   * @throws ParserError if parsing fails
   */
  abstract parse(content: string): Message[];

  /**
   * Validate that content is non-empty.
   *
   * @param content - The content to validate
   * @throws ParserError if content is empty
   */
  protected validateContent(content: string): void {
    if (!content || content.trim().length === 0) {
      throw new ParserError('Log content is empty');
    }
  }
}
